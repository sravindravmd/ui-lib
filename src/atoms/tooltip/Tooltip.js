import React, { useState, useEffect, useRef } from 'react'
import PropTypes from 'prop-types'
import { v4 as uuidv4 } from 'uuid'
import { StyledTooltip, Tooltip_types, Title, Text } from './Tooltip.styles'

const styles = {
  position: 'relative',
  cursor: 'default',
}

const useTooltip = () => {
  const [show, setShow] = useState(false)
  const uid = useRef(uuidv4())

  useEffect(() => {
    const parent = document.getElementById(uid.current).parentElement
    if (parent) {
      parent.style.position = styles.position
      parent.style.cursor = styles.cursor

      const handleMouseEvent = e => {
        return e.type === 'mouseenter' ? setShow(true) : setShow(false)
      }
      parent.addEventListener('mouseenter', handleMouseEvent)
      parent.addEventListener('mouseleave', handleMouseEvent)

      return () => {
        parent.removeEventListener('mouseenter', handleMouseEvent)
        parent.removeEventListener('mouseleave', handleMouseEvent)
      }
    }
  }, [])
  return { uid, show }
}

export const Tooltip = ({
  size,
  position,
  offsetX,
  offsetY,
  title,
  text,
  ...props
}) => {
  const [coordinates, setCoordinates] = useState({ x: 0, y: 0 })
  const { uid, show } = useTooltip()

  const StyledTooltipType = Tooltip_types[position]
    ? Tooltip_types[position]
    : Tooltip_types['bottom']

  useEffect(() => {
    const parent = document.getElementById(uid.current).parentElement
    const handleMousemove = e => {
      setCoordinates({
        x: e.clientX - parent.getBoundingClientRect().x + offsetX,
        y: e.clientY - parent.getBoundingClientRect().y + offsetY,
      })
    }
    if (parent && size === 'small') {
      parent.addEventListener('mousemove', handleMousemove)
      return () => parent.removeEventListener('mousemove', handleMousemove)
    }
  }, [uid, show, offsetX, offsetY])

  return size === 'small' ? (
    <StyledTooltip {...props} id={uid.current} show={show} {...coordinates}>
      {text && <Text>{text}</Text>}
    </StyledTooltip>
  ) : (
    <StyledTooltipType {...props} id={uid.current} show={show}>
      {title && <Title>{title}</Title>}
      {text && <Text>{text}</Text>}
    </StyledTooltipType>
  )
}

Tooltip.defaultProps = {
  position: 'bottom',
  size: 'large',
  offsetX: -5,
  offsetY: 17,
}

Tooltip.propTypes = {
  title: PropTypes.string,
  text: PropTypes.string.isRequired,
  position: PropTypes.oneOf(['right', 'bottom']),
  size: PropTypes.oneOf(['small', 'large']),
  offsetX: PropTypes.number,
  offsetY: PropTypes.number,
}
