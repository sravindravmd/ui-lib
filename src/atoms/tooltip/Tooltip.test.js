import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'
import { Tooltip } from './Tooltip'

afterEach(cleanup)

jest.mock('uuid', () => {
  return {
    v4: jest.fn(() => 1),
  }
})

describe('<Tooltip/>', () => {
  it('renders tooltip with size:large and position:right correctly', () => {
    const { container } = render(
      <Tooltip
        title='Tooltip'
        text='Text for tooltip goes here and it is responsive'
        size='large'
        position='right'
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders tooltip with size:large and position:bottom correctly', () => {
    const { container } = render(
      <Tooltip
        title='Tooltip'
        text='Text for tooltip goes here and it is responsive'
        size='large'
        position='bottom'
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders tooltip with size:small and position at cursor correctly', () => {
    const { container } = render(
      <Tooltip
        title='Tooltip'
        text='Text for tooltip goes here and it is responsive'
        size='small'
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('show tooltip on mouse enter event', () => {
    const { container, getByText } = render(
      <Tooltip title='Tooltip' text='Text for tooltip' size='large' />
    )
    fireEvent.mouseEnter(container)
    const tooltip = getByText('Tooltip')
    expect(tooltip).toBeInTheDocument()
  })

  it('show small tooltip on mouse move event', () => {
    const { container, getByText } = render(
      <Tooltip text='Text for tooltip' size='small' />
    )
    fireEvent.mouseEnter(container)
    const tooltip = getByText('Text for tooltip')
    expect(tooltip).toBeInTheDocument()
  })
})
