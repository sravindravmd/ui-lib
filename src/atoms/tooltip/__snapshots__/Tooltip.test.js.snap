// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip/> renders tooltip with size:large and position:bottom correctly 1`] = `
<div
  style="position: relative; cursor: default;"
>
  <div
    class="sc-bdfBwQ sc-jSgupP eBdUkv iskboO"
    id="1"
  >
    <div
      class="sc-dlfnbm kTihnT"
    >
      Tooltip
    </div>
    <div
      class="sc-hKgILt eWZXxS"
    >
      Text for tooltip goes here and it is responsive
    </div>
  </div>
</div>
`;

exports[`<Tooltip/> renders tooltip with size:large and position:right correctly 1`] = `
<div
  style="position: relative; cursor: default;"
>
  <div
    class="sc-bdfBwQ sc-eCssSg eBdUkv fHyGHs"
    id="1"
  >
    <div
      class="sc-dlfnbm kTihnT"
    >
      Tooltip
    </div>
    <div
      class="sc-hKgILt eWZXxS"
    >
      Text for tooltip goes here and it is responsive
    </div>
  </div>
</div>
`;

exports[`<Tooltip/> renders tooltip with size:small and position at cursor correctly 1`] = `
<div
  style="position: relative; cursor: default;"
>
  <div
    class="sc-bdfBwQ sc-gsTCUz eBdUkv gfvFja"
    id="1"
    x="0"
    y="0"
  >
    <div
      class="sc-hKgILt eWZXxS"
    >
      Text for tooltip goes here and it is responsive
    </div>
  </div>
</div>
`;
